test/common/network/udp_listener_impl_test.cc

```diff
@@ -505,6 +423,73 @@ TEST_P(UdpListenerImplTest, UdpListenerRecvFromError) {
   dispatcher_->run(Event::Dispatcher::RunType::Block);
 }
 
+/**
+ * Tests UDP listener for sending datagrams to destination.
+ *  1. Setup a server udp listener and client udp listener
+ *  2. Send the data from the client component to the server and validate the contents
+ */
+TEST_P(UdpListenerImplTest, SendData) {
+  // Setup server socket
+  SocketPtr server_socket =
+      getSocket(Address::SocketType::Datagram, Network::Test::getCanonicalLoopbackAddress(version_),
+                nullptr, true);
+
+  ASSERT_NE(server_socket, nullptr);
+
+  auto const* server_ip = server_socket->localAddress()->ip();
+  ASSERT_NE(server_ip, nullptr);
+
+  // Setup server callback handler and listener.
+  Network::MockUdpListenerCallbacks server_listener_callbacks;
+  Network::TestUdpListenerImpl server_listener(dispatcherImpl(), *server_socket.get(),
+                                               server_listener_callbacks);
+
+  EXPECT_CALL(server_listener, doRecvFrom(_, _))
+      .WillRepeatedly(Invoke([&](sockaddr_storage& peer_addr, socklen_t& addr_len) {
+        return server_listener.doRecvFrom_(peer_addr, addr_len);
+      }));
+
+  // Setup client socket.
+  SocketPtr client_socket =
+      getSocket(Address::SocketType::Datagram, Network::Test::getCanonicalLoopbackAddress(version_),
+                nullptr, false);
+
+  ASSERT_NE(client_socket, nullptr);
+
+  // Setup client callback handler and listener.
+  Network::MockUdpListenerCallbacks client_listener_callbacks;
+  Network::TestUdpListenerImpl client_listener(dispatcherImpl(), *client_socket.get(),
+                                               client_listener_callbacks);
+
+  const std::string payload("hello world");
+  Buffer::InstancePtr buffer(new Buffer::OwnedImpl());
+  buffer->add(payload);
+  UdpSendData send_data{server_socket->localAddress(), *buffer};
+
+  client_listener.send(send_data);
```

Comment:
The fact that client has a listener is a bit weird to me. Can you just use client_socket->io_handle() to send test data?

Reply:
I can rename the variable names (to 2 servers rather than 1 client and 1 server) - but the intention here is to exercise the "send" method on the udp listener and get that data on the other side and validate its contents.

2 instances of the listener are hence used to do the above - 1 for testing the send code path and the other for receiving it and validating the contents.

Reply:
io_handle().recvFrom() is already tested in other tests. I suggested to omit client listener just for code simplicity. You can call server_listener.send() and then client_socket->io_handle().recvfrom() to test listener sending data.

Reply:
ah, I see your point. Sure, that can be done